src/Opaleye/Internal/RunQuery.hs

```diff
@@ -327,23 +353,51 @@ jsonFieldParser, jsonbFieldParser :: FieldParser String
 jsonFieldParser  = jsonFieldTypeParser (String.fromString "json")
 jsonbFieldParser = jsonFieldTypeParser (String.fromString "jsonb")
 
+jsonFieldTextParser, jsonbFieldTextParser :: FieldParser ST.Text
+jsonFieldTextParser  = jsonFieldTypeTextParser (String.fromString "json")
+jsonbFieldTextParser = jsonFieldTypeTextParser (String.fromString "jsonb")
+
+jsonFieldLazyTextParser, jsonbFieldLazyTextParser :: FieldParser LT.Text
+jsonFieldLazyTextParser  = jsonFieldTypeLazyTextParser (String.fromString "json")
+jsonbFieldLazyTextParser = jsonFieldTypeLazyTextParser (String.fromString "jsonb")
+
+jsonFieldByteParser, jsonbFieldByteParser :: FieldParser SBS.ByteString
+jsonFieldByteParser  = jsonFieldTypeByteParser (String.fromString "json")
+jsonbFieldByteParser = jsonFieldTypeByteParser (String.fromString "jsonb")
+
+jsonFieldLazyByteParser, jsonbFieldLazyByteParser :: FieldParser LBS.ByteString
+jsonFieldLazyByteParser  = jsonFieldTypeLazyByteParser (String.fromString "json")
+jsonbFieldLazyByteParser = jsonFieldTypeLazyByteParser (String.fromString "jsonb")
+
 -- typenames, not type Oids are used in order to avoid creating
 -- a dependency on 'Database.PostgreSQL.LibPQ'
 --
 -- Eventually we want to move this to postgresql-simple
 --
 --     https://github.com/tomjaguarpaw/haskell-opaleye/issues/329
 jsonFieldTypeParser :: SBS.ByteString -> FieldParser String
-jsonFieldTypeParser jsonTypeName field mData = do
+jsonFieldTypeParser x = (fmap . fmap . fmap) IPT.strictDecodeUtf8 (jsonFieldTypeByteParser x)
+
+jsonFieldTypeTextParser :: SBS.ByteString -> FieldParser ST.Text
+jsonFieldTypeTextParser x = (fmap . fmap . fmap) STE.decodeUtf8 (jsonFieldTypeByteParser x)
+
+jsonFieldTypeLazyTextParser :: SBS.ByteString -> FieldParser LT.Text
+jsonFieldTypeLazyTextParser x = (fmap . fmap . fmap) (LTE.decodeUtf8 . LBS.fromStrict) (jsonFieldTypeByteParser x)
+
+jsonFieldTypeByteParser :: SBS.ByteString -> FieldParser SBS.ByteString
+jsonFieldTypeByteParser jsonTypeName field mData = do
     ti <- typeInfo field
     if TI.typname ti == jsonTypeName
        then convert
        else returnError Incompatible field "types incompatible"
   where
     convert = case mData of
-        Just bs -> pure $ IPT.strictDecodeUtf8 bs
+        Just bs -> pure bs
         _       -> returnError UnexpectedNull field ""
 
+jsonFieldTypeLazyByteParser :: SBS.ByteString -> FieldParser LBS.ByteString
+jsonFieldTypeLazyByteParser x = (fmap . fmap . fmap) LBS.fromStrict (jsonFieldTypeByteParser x)
+
```

Comment:
Nice! Do you mind adding another `fmap` and getting rid of those `x`s? :D

Reply:
Done 😄